spark-cassandra-connector/src/main/scala/org/apache/spark/sql/cassandra/DefaultSource.scala

```diff
@@ -0,0 +1,93 @@
+  override def createRelation(sqlContext: SQLContext,
+                              parameters: Map[String, String]): BaseRelation = {
+    val schemaJsonString = getUserDefinedSchema(parameters)
+    val schema = if (schemaJsonString == None) None
+    else Option(DataType.fromJson(schemaJsonString.get).asInstanceOf[StructType])
+    sqlContext.getDataSourceRelation(getTable(parameters), getKeyspace(parameters),
+      getScanType(parameters), getCluster(parameters), schema)
+  }
+
+  /** Creates a new relation for a cassandra table given table, keyspace, cluster and schema as a parameter.*/
+  override def createRelation(sqlContext: SQLContext,
+                              parameters: Map[String, String],
+                              schema: StructType): BaseRelation = {
+    sqlContext.getDataSourceRelation(getTable(parameters), getKeyspace(parameters),
+      getScanType(parameters), getCluster(parameters), Option(schema))
+  }
+
+
+  /** Creates a new relation for a cassandra table given table, keyspace, cluster, schema and data as a parameter.*/
+  override def createRelation(sqlContext: SQLContext,
+                              mode: SaveMode,
+                              parameters: Map[String, String],
+                              data: DataFrame): BaseRelation = {
+    val table = sqlContext.getDataSourceRelation(getTable(parameters), getKeyspace(parameters),
+      getScanType(parameters), getCluster(parameters), None)
+
+    mode match {
+      case Append => table.insert(data, false)
+      case Overwrite => throw new UnsupportedOperationException("Overwriting a Cassandra Table is not allowed.")
+      case ErrorIfExists =>
+        if (!table.buildScan().isEmpty()) {
+          throw new UnsupportedOperationException("Writing to a none-empty Cassandra Table is not allowed.")
+        } else {
+          table.insert(data, false)
+        }
+      case Ignore =>
+    }
+
+    sqlContext.getDataSourceRelation(getTable(parameters), getKeyspace(parameters),
+      getScanType(parameters), getCluster(parameters), None)
+  }
+
+
+  import DefaultSource._
+
+  private def getTable(parameters: Map[String, String]) : String =
+    parameters.getOrElse(CassandraDataSourceTableNameProperty, missingProp(CassandraDataSourceTableNameProperty))
+
+  private def getKeyspace(parameters: Map[String, String]) : String =
+    parameters.getOrElse(CassandraDataSourceKeyspaceNameProperty, missingProp(CassandraDataSourceKeyspaceNameProperty))
+
+  private def getCluster(parameters: Map[String, String]): Option[String] =
+    parameters.get(CassandraDataSourceClusterNameProperty)
+
+  private def getUserDefinedSchema(parameters: Map[String, String]) : Option[String] =
+    parameters.get(CassandraDataSourUserDefinedSchemaNameProperty)
+
+  private def missingProp(prop: String) = throw new IllegalArgumentException(s"Missing $prop name")
+
+  private def getScanType(parameters: Map[String, String]) : ScanType = CassandraDataSourceScanTypeMap.get(
```

Comment:
This can be simplified and flattened.

```scala
  private def scanTypeName(parameters: Map[String, String]): String = 
    parameters.getOrElse(
      CassandraDataSourceScanTypeNameProperty, 
      CassandraDataSourcePrunedFilteredScanTypeName)
  
  private def getScanType(parameters: Map[String, String]) : ScanType = {
    val stn = scanTypeName(parameters).toLowerCase
    CassandraDataSourceScanTypeMap.getOrElse(stn, CatalystScanType)
  }
}
```